File: apps/desktop/crates/domain/src/filters.rs
//! Issue filters — a VERBATIM port of `apps/web/src/lib/filters.ts`
//! (masterplan-v3 §4.7).
//!
//! The `IssueFilters` shape and `matches_filters()` are mirrored across four
//! clients: web (`lib/filters.ts`), iOS (`Domain/IssueFilters.swift`), Android
//! (`domain/IssueFilters.kt`) and this crate. If you change the filter shape
//! or matching semantics here, update the other three to keep the clients in
//! lockstep (no shared package yet).

use crate::enums::IssuePriority;
use crate::rows::Issue;
use crate::statuses::{status_key_matches, ResolvedStatus};

/// `IssueFilters` — web `interface IssueFilters`.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct IssueFilters {
    /// EXP-314: RESOLVED status group keys (an `issue_statuses` row id, or
    /// `builtin:<key>` while the statuses shape has not synced) — not enum
    /// values. An issue matches when its resolved group key is in the set.
    pub status_keys: Vec<String>,
    pub priorities: Vec<IssuePriority>,
    pub label_ids: Vec<String>,
}

/// Web `emptyFilters`.
pub fn empty_filters() -> IssueFilters {
    IssueFilters::default()
}

impl IssueFilters {
    /// Web `emptyFilters` as an associated constructor.
    pub fn empty() -> Self {
        Self::default()
    }
}

/// Web `matchesFilters(issue, issueLabelIds, filters, resolvedStatus)`. Each
/// active category must match (AND across categories); within a category any
/// value matches (OR). An empty category is a pass. `status` is the issue's
/// RESOLVED status (`domain::statuses::resolve_status`) — matching goes
/// through [`status_key_matches`], so a `builtin:<key>` token stored before
/// the statuses shape synced keeps selecting the synced row it re-keyed to.
pub fn matches_filters(
    issue: &Issue,
    issue_label_ids: &[String],
    status: &ResolvedStatus,
    filters: &IssueFilters,
) -> bool {
    if !filters.status_keys.is_empty()
        && !filters
            .status_keys
            .iter()
            .any(|token| status_key_matches(status, token))
    {
        return false;
    }
    if !filters.priorities.is_empty() && !filters.priorities.contains(&issue.priority) {
        return false;
    }
    if !filters.label_ids.is_empty()
        && !filters
            .label_ids
            .iter()
            .any(|id| issue_label_ids.contains(id))
    {
        return false;
    }
    true
}

/// Web `activeFilterCount(filters)`.
pub fn active_filter_count(filters: &IssueFilters) -> usize {
    filters.status_keys.len() + filters.priorities.len() + filters.label_ids.len()
}

/// Web `hasActiveFilters(filters)`.
pub fn has_active_filters(filters: &IssueFilters) -> bool {
    active_filter_count(filters) > 0
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::options::{ColorToken, IconGlyph};
    use crate::statuses::{IssueStatusCategory, StatusTint};
    use serde_json::json;

    /// A synced status row as the UI resolved it (`group_key` = row uuid).
    fn synced(group_key: &str, builtin_key: Option<&str>) -> ResolvedStatus {
        ResolvedStatus {
            group_key: group_key.to_string(),
            row_id: Some(group_key.to_string()),
            name: group_key.to_string(),
            category: IssueStatusCategory::Unstarted,
            tint: StatusTint::Token(ColorToken::Foreground),
            glyph: IconGlyph::Circle,
            builtin_key: builtin_key.map(str::to_string),
        }
    }

    /// A CONSTRUCTED fallback (`group_key` = `builtin:<key>`).
    fn fallback(builtin_key: &str) -> ResolvedStatus {
        ResolvedStatus {
            group_key: format!("builtin:{builtin_key}"),
            row_id: None,
            ..synced(builtin_key, Some(builtin_key))
        }
    }

    fn issue(status: &str, priority: &str) -> Issue {
        serde_json::from_value(json!({
            "id": "i-1",
            "board_id": "p-1",
            "number": 1,
            "identifier": "EXP-1",
            "title": "t",
            "status": status,
            "priority": priority
        }))
        .unwrap()
    }

    #[test]
    fn empty_filters_is_empty() {
        let filters = empty_filters();
        assert!(filters.status_keys.is_empty());
        assert!(filters.priorities.is_empty());
        assert!(filters.label_ids.is_empty());
        assert_eq!(filters, IssueFilters::empty());
    }

    #[test]
    fn matches_filters_passes_everything_when_empty() {
        let filters = empty_filters();
        assert!(matches_filters(&issue("backlog", "none"), &[], &synced("s-1", None), &filters));
        assert!(matches_filters(
            &issue("done", "urgent"),
            &["l-1".to_string()],
            &fallback("done"),
            &filters
        ));
    }

    #[test]
    fn matches_filters_status_category_is_group_keys() {
        let filters = IssueFilters {
            status_keys: vec!["s-backlog".to_string(), "s-qa".to_string()],
            ..Default::default()
        };
        assert!(matches_filters(
            &issue("backlog", "none"),
            &[],
            &synced("s-backlog", Some("backlog")),
            &filters
        ));
        // A custom started status matches by KEY, not by its `in_progress`
        // anchor.
        assert!(matches_filters(
            &issue("in_progress", "none"),
            &[],
            &synced("s-qa", None),
            &filters
        ));
        assert!(!matches_filters(
            &issue("in_progress", "none"),
            &[],
            &synced("s-in-progress", Some("in_progress")),
            &filters
        ));
        // Pre-sync fallback keys work the same way.
        let filters = IssueFilters {
            status_keys: vec!["builtin:backlog".to_string()],
            ..Default::default()
        };
        assert!(matches_filters(
            &issue("backlog", "none"),
            &[],
            &fallback("backlog"),
            &filters
        ));
    }

    #[test]
    fn a_builtin_token_still_matches_the_synced_row_it_rekeyed_to() {
        // Stored while only the constructed vocabulary existed; the shape has
        // since synced and the group key is the row uuid now.
        let filters = IssueFilters {
            status_keys: vec!["builtin:in_progress".to_string()],
            ..Default::default()
        };
        assert!(matches_filters(
            &issue("in_progress", "none"),
            &[],
            &synced("row-wip", Some("in_progress")),
            &filters
        ));
        // …but it must not select a different builtin, nor a custom row.
        assert!(!matches_filters(
            &issue("backlog", "none"),
            &[],
            &synced("row-backlog", Some("backlog")),
            &filters
        ));
        assert!(!matches_filters(
            &issue("in_progress", "none"),
            &[],
            &synced("row-qa", None),
            &filters
        ));
        // A row-uuid token keeps matching exactly one row.
        let filters = IssueFilters {
            status_keys: vec!["row-wip".to_string()],
            ..Default::default()
        };
        assert!(matches_filters(
            &issue("in_progress", "none"),
            &[],
            &synced("row-wip", Some("in_progress")),
            &filters
        ));
        assert!(!matches_filters(
            &issue("in_progress", "none"),
            &[],
            &fallback("in_progress"),
            &filters
        ));
    }

    #[test]
    fn matches_filters_priority_category() {
        let filters = IssueFilters {
            priorities: vec![IssuePriority::Urgent],
            ..Default::default()
        };
        assert!(matches_filters(
            &issue("backlog", "urgent"),
            &[],
            &synced("s-1", None),
            &filters
        ));
        assert!(!matches_filters(&issue("backlog", "low"), &[], &synced("s-1", None), &filters));
    }

    #[test]
    fn matches_filters_labels_are_any_of() {
        let filters = IssueFilters {
            label_ids: vec!["l-1".to_string(), "l-2".to_string()],
            ..Default::default()
        };
        // web: filters.labelIds.some((id) => issueLabelIds.includes(id))
        assert!(matches_filters(
            &issue("backlog", "none"),
            &["l-2".to_string(), "l-9".to_string()],
            &synced("s-1", None),
            &filters
        ));
        assert!(!matches_filters(
            &issue("backlog", "none"),
            &["l-9".to_string()],
            &synced("s-1", None),
            &filters
        ));
        assert!(!matches_filters(&issue("backlog", "none"), &[], &synced("s-1", None), &filters));
    }

    #[test]
    fn matches_filters_is_and_across_categories() {
        let filters = IssueFilters {
            status_keys: vec!["s-backlog".to_string()],
            priorities: vec![IssuePriority::High],
            label_ids: vec!["l-1".to_string()],
        };
        assert!(matches_filters(
            &issue("backlog", "high"),
            &["l-1".to_string()],
            &synced("s-backlog", Some("backlog")),
            &filters
        ));
        // Right status + label, wrong priority → fail.
        assert!(!matches_filters(
            &issue("backlog", "low"),
            &["l-1".to_string()],
            &synced("s-backlog", Some("backlog")),
            &filters
        ));
    }

    #[test]
    fn active_filter_count_sums_all_categories() {
        let filters = IssueFilters {
            status_keys: vec!["s-backlog".to_string(), "s-done".to_string()],
            priorities: vec![IssuePriority::Low],
            label_ids: vec!["l-1".to_string()],
        };
        assert_eq!(active_filter_count(&filters), 4);
        assert!(has_active_filters(&filters));
        assert_eq!(active_filter_count(&empty_filters()), 0);
        assert!(!has_active_filters(&empty_filters()));
    }
}

File: apps/web/src/lib/filters.ts
// The IssueFilters shape and matchesFilters() are mirrored across three
// clients. If you change the filter shape or matching semantics here, also
// update apps/ios/Exponential/Domain/IssueFilters.swift and
// apps/android/app/src/main/java/com/exponential/app/domain/IssueFilters.kt
// to keep the three clients in sync (no shared package yet).
import type { Issue } from "@/db/schema"
import type { IssuePriority } from "@/lib/domain"
import { issuePriorityOptions, issueStatusOptions } from "@/lib/domain"

export interface IssueFilters {
  // EXP-314 dual-token status filter: a token is either an `issue_statuses`
  // row uuid (the modern form the picker writes) or a legacy anchor-enum value
  // (`backlog`, `done`, …) carried by an older shared/bookmarked URL or an older
  // native client. Matching accepts both; the picker normalizes to row ids on
  // the next toggle.
  statusTokens: string[]
  priorities: IssuePriority[]
  labelIds: string[]
}

export const emptyFilters: IssueFilters = {
  statusTokens: [],
  priorities: [],
  labelIds: [],
}

export function matchesFilters(
  issue: Issue,
  issueLabelIds: string[],
  filters: IssueFilters,
  // The issue's RESOLVED status (from useTeamStatuses().resolve) — matching
  // on it keeps filters agreeing with the rendered groups even for rows whose
  // status_id is NULL/stale (pre-backfill, deleted-status races), exactly
  // like the three native mirrors. Callers without team rows omit it and get
  // the raw dual-column match.
  resolvedStatus?: { id: string; builtinKey: string | null }
): boolean {
  if (
    filters.statusTokens.length > 0 &&
    !filters.statusTokens.some((token) =>
      resolvedStatus
        ? token === resolvedStatus.id || token === resolvedStatus.builtinKey
        : // A uuid token only ever equals `statusId`; an enum token only ever
          // equals `status` — one dual comparison covers both without team
          // data.
          token === issue.statusId || token === issue.status
    )
  )
    return false
  if (
    filters.priorities.length > 0 &&
    !filters.priorities.includes(issue.priority)
  )
    return false
  if (
    filters.labelIds.length > 0 &&
    !filters.labelIds.some((id) => issueLabelIds.includes(id))
  )
    return false
  return true
}

export function activeFilterCount(filters: IssueFilters): number {
  return (
    filters.statusTokens.length +
    filters.priorities.length +
    filters.labelIds.length
  )
}

export function hasActiveFilters(filters: IssueFilters): boolean {
  return activeFilterCount(filters) > 0
}

// --- URL search-param form (web-only, not mirrored on native) ---------------
// Filters live in the URL as clean comma-joined values (?status=done,in_progress
// &priority=high&labels=<id>,<id>) so a filtered board is shareable and survives
// a refresh. The issue-detail route accepts the SAME optional params so the
// prev/next switcher can follow the board's filtered ordering; helpers are
// shared here so the two routes can't drift.
//
// EXP-314 keeps the `status` KEY and widens its allowlist to row uuids — old
// links (?status=done) keep working verbatim.

export interface IssueFilterSearch {
  status?: string
  priority?: string
  labels?: string
}

const STATUS_VALUES: readonly string[] = issueStatusOptions.map((o) => o.value)
const PRIORITY_VALUES: readonly string[] = issuePriorityOptions.map(
  (o) => o.value
)
const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i

export function isValidStatusToken(value: string): boolean {
  return UUID_RE.test(value) || STATUS_VALUES.includes(value)
}

// Coerce a raw search value (array or comma string) to a validated, comma-joined
// string, or undefined when empty — so cleared filters drop out of the URL.
function validatedCsv(
  raw: unknown,
  allowed?: (value: string) => boolean
): string | undefined {
  let arr: string[]
  if (Array.isArray(raw)) {
    arr = raw.filter((v): v is string => typeof v === `string`)
  } else if (typeof raw === `string` && raw.length > 0) {
    arr = raw.split(`,`)
  } else {
    return undefined
  }
  const cleaned = allowed
    ? arr.filter((v) => allowed(v))
    : arr.filter((v) => v.length > 0)
  return cleaned.length ? cleaned.join(`,`) : undefined
}

const isKnownPriority = (value: string) => PRIORITY_VALUES.includes(value)

// validateSearch body for any route carrying board filters; drops anything
// unrecognised.
export function parseIssueFilterSearch(
  search: Record<string, unknown>
): IssueFilterSearch {
  return {
    status: validatedCsv(search.status, isValidStatusToken),
    priority: validatedCsv(search.priority, isKnownPriority),
    labels: validatedCsv(search.labels),
  }
}

export function issueFiltersFromSearch(search: IssueFilterSearch): IssueFilters {
  return {
    statusTokens: search.status ? search.status.split(`,`) : [],
    priorities: search.priority
      ? (search.priority.split(`,`) as IssuePriority[])
      : [],
    labelIds: search.labels ? search.labels.split(`,`) : [],
  }
}

// Always emits all three keys (undefined when empty) so spreading over `prev`
// search state clears removed filters from the URL.
export function issueFilterSearchFromFilters(
  filters: IssueFilters
): IssueFilterSearch {
  return {
    status: filters.statusTokens.length
      ? filters.statusTokens.join(`,`)
      : undefined,
    priority: filters.priorities.length
      ? filters.priorities.join(`,`)
      : undefined,
    labels: filters.labelIds.length ? filters.labelIds.join(`,`) : undefined,
  }
}
